### PR TITLE
added new method to load environment from URL and added UTs

### PR DESCRIPTION
**Pull request by seyadava**
Thank you for your contribution to Go-AutoRest! We will triage and review it as soon as we can.

As part of submitting, please make sure you can make the following assertions:
 - [x] I've tested my changes, adding unit tests if applicable.
 - [x] I've added Apache 2.0 Headers to the top of any new source files.
 - [x] I'm submitting this PR to the `dev` branch, except in the case of urgent bug fixes warranting their own release.
 - [x] If I'm targeting `master`, I've updated [CHANGELOG.md](https://github.com/Azure/go-autorest/blob/master/CHANGELOG.md) to address the changes I'm making.

### CODE REVIEW

File: autorest/azure/metadata_environment.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+// OverrideProperty represents property name and value that clients can override
+type OverrideProperty struct {
+	Key   EnvironmentProperty
+	Value string
+}
+
+// EnvironmentFromURL loads an Environment from a URL
+// This function is particularly useful in the Hybrid Cloud model, where one may define their own
+// endpoints.
+func EnvironmentFromURL(endpoint string, properties ...OverrideProperty) (environment Environment, err error) {
```
Comment (bganapa) [inline autorest/azure/metadata_environment.go]: rename endpoint to ArmEndpoint or resourceManagerEndpoint

File: autorest/azure/environments.go [hunk 2]
```diff
@@ -78,6 +79,7 @@ var (
 		ServiceManagementVMDNSSuffix: "cloudapp.net",
 		ResourceManagerVMDNSSuffix:   "cloudapp.azure.com",
 		ContainerRegistryDNSSuffix:   "azurecr.io",
+		TokenAudience:                "https://management.azure.com/",
```
Comment (jhendrixMSFT) [inline autorest/azure/environments.go]: This is a duplicate of ```ResourceManagerEndpoint```.  Why is this necessary?

File: autorest/azure/testdata/test_metadata_environment_1.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{
+    "galleryEndpoint":"https://portal.local.azurestack.external:30015/",
+    "graphEndpoint":"https://graph.windows.net/",
+    "portalEndpoint":"https://portal.local.azurestack.external/",
+    "authentication":{
+        "loginEndpoint":"https://login.windows.net/",
+        "audiences":["https://management.azurestackci04.onmicrosoft.com/3ee899c8-c137-4ce4-b230-619961a09a73"]
+    }
+}
```
Comment (jhendrixMSFT) [inline autorest/azure/testdata/test_metadata_environment_1.json]: Missing newline.

File: autorest/azure/metadata_environment.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+	jsonResponse, err := ioutil.ReadAll(response.Body)
+	err = json.Unmarshal(jsonResponse, &environment)
+	return
+}
+
+func getStampDNSSuffix(portalEndpoint string) string {
+	return strings.TrimSuffix(strings.TrimPrefix(strings.Replace(portalEndpoint, strings.Split(portalEndpoint, ".")[0], "", 1), "."), "/")
+}
+
+func userAgent() string {
```
Comment (jhendrixMSFT) [inline autorest/azure/metadata_environment.go]: Can we use ```defaultUserAgent``` instead?  See [here](https://github.com/Azure/go-autorest/blob/19642ad23ee58fb8c895dfdc97c979c15c133118/autorest/client.go#L46).

File: autorest/azure/metadata_environment.go [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+	if err != nil {
+		return environment, err
+	}
+	defer response.Body.Close()
+	jsonResponse, err := ioutil.ReadAll(response.Body)
+	err = json.Unmarshal(jsonResponse, &environment)
+	return
+}
+
+func getStampDNSSuffix(portalEndpoint string) string {
```
Comment (jhendrixMSFT) [inline autorest/azure/metadata_environment.go]: It looks like this is called in only one place, can it be folded into the caller?

File: autorest/azure/environments.go (hunk 2 above)
Comment (bganapa) [reply autorest/azure/environments.go]: For AzureStack, the TokenAudience is different from the ResourceManagerEndpoint, Hence this is needed to support AzureStack better

Comment (jhendrixMSFT) [reply autorest/azure/environments.go]: I see, it just defaults to the same value as ResourceManagerEndpoint?

Comment (bganapa) [reply autorest/azure/environments.go]: Yes, in case of AzureClouds audiences value is same as RM endpoint e.g. 
pls look at audiences value
https://management.azure.com/metadata/endpoints?api-version=1.0
https://management.usgovcloudapi.net/metadata/endpoints?api-version=1.0

in case of azure stack, the audiences value differs with each deployment. (i am not sure whether you have access to the following endpoints)
https://management.redmond.azurestack.corp.microsoft.com/metadata/endpoints?api-version=1.0
https://management.shanghai.azurestack.corp.microsoft.com/metadata/endpoints?api-version=1.0

File: autorest/azure/metadata_environment.go (hunk 4 above)
Comment (seyadava) [reply autorest/azure/metadata_environment.go]: defaultUserAgent is not an exposed variable, so we cannot use it to instantiate client. 
However, instead of using NewClientWithUserAgent method, I could directly instantiate a client, like: 
client := &autorest.Client{
		PollingDelay:    autorest.DefaultPollingDelay,
		PollingDuration: autorest.DefaultPollingDuration,
		RetryAttempts:   autorest.DefaultRetryAttempts,
		RetryDuration:   autorest.DefaultRetryDuration,
	}
This way we don't need to pass userAgent. Thoughts?

File: autorest/azure/metadata_environment.go (hunk 1 above)
Comment (seyadava) [reply autorest/azure/metadata_environment.go]: Renamed it to "resourceManagerEndpoint"

File: autorest/azure/testdata/test_metadata_environment_1.json (hunk 3 above)
Comment (seyadava) [reply autorest/azure/testdata/test_metadata_environment_1.json]: fixed.

File: autorest/azure/metadata_environment.go (hunk 5 above)
Comment (seyadava) [reply autorest/azure/metadata_environment.go]: Folded it into the caller

File: autorest/azure/metadata_environment.go (hunk 4 above)
Comment (jhendrixMSFT) [reply autorest/azure/metadata_environment.go]: I think you can continue to use ```NewClientWithUserAgent```, just pass the empty string for the param.

Comment (seyadava) [reply autorest/azure/metadata_environment.go]: Done.

File: autorest/azure/metadata_environment.go [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+func retrieveMetadataEnvironment(endpoint string) (environment environmentMetadataInfo, err error) {
+	client := autorest.NewClientWithUserAgent("")
+	managementEndpoint := fmt.Sprintf("%s%s", strings.TrimSuffix(endpoint, "/"), "/metadata/endpoints?api-version=1.0")
+	req, _ := http.NewRequest("GET", managementEndpoint, nil)
+	response, err := client.Do(req)
+	if err != nil {
+		return environment, err
+	}
+	defer response.Body.Close()
+	jsonResponse, err := ioutil.ReadAll(response.Body)
```
Comment (jhendrixMSFT) [inline autorest/azure/metadata_environment.go]: Can you please update this to return if err is not nil?

File: autorest/azure/metadata_environment.go [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+// OverrideProperty represents property name and value that clients can override
+type OverrideProperty struct {
+	Key   EnvironmentProperty
+	Value string
+}
+
+// EnvironmentFromURL loads an Environment from a URL
+// This function is particularly useful in the Hybrid Cloud model, where one may define their own
+// endpoints.
+func EnvironmentFromURL(resourceManagerEndpoint string, properties ...OverrideProperty) (environment Environment, err error) {
```
Comment (jhendrixMSFT) [inline autorest/azure/metadata_environment.go]: Since this is variadic the caller can elect to omit this parameter entirely.  What happens in that case, should it return an error?

Comment (seyadava) [reply autorest/azure/metadata_environment.go]: If OverrideProperty is not passed, it means caller does not want to override any environment value. In this case method uses default values for all of the environment properties and no error should be returned.
